staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -251,90 +232,95 @@ func (t *tracker) Add(obj runtime.Object) error {
 	if meta.IsListType(obj) {
 		return t.addList(obj, false)
 	}
-
 	objMeta, err := meta.Accessor(obj)
 	if err != nil {
 		return err
 	}
-	return t.add(obj, objMeta.GetNamespace(), false)
-}
-
-func (t *tracker) Create(obj runtime.Object, ns string) error {
-	return t.add(obj, ns, false)
-}
-
-func (t *tracker) Update(obj runtime.Object, ns string) error {
-	return t.add(obj, ns, true)
-}
-
-func (t *tracker) add(obj runtime.Object, ns string, replaceExisting bool) error {
 	gvks, _, err := t.scheme.ObjectKinds(obj)
 	if err != nil {
 		return err
 	}
 	if len(gvks) == 0 {
 		return fmt.Errorf("no registered kinds for %v", obj)
 	}
-
-	t.lock.Lock()
-	defer t.lock.Unlock()
-
 	for _, gvk := range gvks {
-		gr := schema.GroupResource{Group: gvk.Group, Resource: gvk.Kind}
+		// NOTE: KindToResource is a heuristic and default match. The actual
+		// registration in apiserver can specify arbitrary route for a gvk. If a
+		// test uses such objects, it cannot preset the tracker with objects via
+		// Add(). Instead, it should trigger the Create() function of the
+		// tracker, where an arbitrary gvr can be specified.
+		gvr, _ := meta.KindToResource(gvk)
+		// Resource doesn't have the concept of "__internal" version, just set it to "".
+		if gvr.Version == runtime.APIVersionInternal {
```

Comment:
That means we need to change hundreds lines of manually written test code like this [one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller_test.go#L163). It also requires changing the generated fake clients.

Apart from the massive changes, i don't feel right writing `schema.GroupVersionResource{Group:"core, Version:"__internal", Resource: "Pod"}` in tests because there is no such GVR in production code.